config/master.key

```diff
@@ -0,0 +1 @@
+9cbfdc66d80a5b532b1b14c0e0cf16ff
```

Comment:
While this is just a randomly generated key, officially the credential keys are not supposed to be committed to version control.  However, I have been unable to find a way to make this work "out of the box" (at least with s2i) without this being present.  Any better ideas?

Reply:
Well...
```
The next important file is config/master.key which is where your RAILS_MASTER_KEY will be kept. Now this is a very important file that can never be committed to your source control tool! I can’t stress this enough because this key will give anyone access to your encrypted keys and will leave your application vulnerable to attacks. To avoid this you need to add your config/master.key to your .gitignore file and this will keep your master.key a secret.
```

Not having it would mean we need to regenerate it on every s2i build (or users need to do it).

```
$ bin/rails credentials:edit
```
```
One thing to know is that if you don’t have either of these files, running the above command will automatically create them for you (thanks Rails).
```

Or it can be stored in some OpenShift secrets / ENV:
```
Option 2: create a RAILS_MASTER_KEY ENV variable. Rails will detect it and use it as your master key, e.g. in Heroku: heroku config:set RAILS_MASTER_KEY=<your-master-key-here>.
```


https://medium.com/@thorntonbrenden/rails-and-the-legendary-master-key-15c8be7799f1

Reply:
@yselkowitz @jackorp Do you want to try to fix this?

Reply:
That seems doable for the openshift template, except that you still have to have something for 1) pure s2i, 2) those using previous versions of the templates (e.g. all current versions of OpenShift).  What's the intended fallback then?

Reply:
Also, iiuc the master key is used to encrypt credentials.yml.enc, so if we want to introduce a RAILS_MASTER_KEY env, then afaics there has to be the means to create credentials during s2i/assemble (and also non-interactively, as `rails credentials:edit` uses $EDITOR).  

FWIW, django-ex provides a [fallback secret key](https://github.com/sclorg/django-ex/blob/3.2.x/project/settings.py#L27) but I suppose it doesn't have the equivalent of credentials.yml.enc?

Reply:
I think there are two options:

 - Support the deployment using OpenShift (latest), where you'd need to prepare the template to request / generate some key for you, and save it as secret. 
   -  _Older versions of OpenShift would fail_ ... So either new repo needs to be created or we need to point to some branch.
   - _s2i wouldn't work?_  ... Well, you can have s2i ENV as well, defined by user - this needs to be documented at relevant places. I don't think there's any technical difference from OpenShift here.
   - _Alternatively_ ... We can code some default / fallback key there. It shouldn't be hard, but then again we're at **insecure master key** again (or the key will be inaccessible by the user.
 - Leave the app as it is, with the insecure setup.
   - _Stress for users that it's not for production purposes_ ... We can put the message on the main page of the test app f.e..
   - _Additionally, we can have some guide, explaining:_ ... For any setup where the app would be accessible on the internet / used for more than testing there would need to be the key regenerated (by cloning/editing the git and redeployment).

Reply:
Is there a third option?
- Merge this as-is with the insecure key as an out-of-the-box fallback, then separately implement a RAILS_MASTER_KEY environment variable(s) in the template along with changes to ruby's s2i/assemble to run (where appropriate) something like `EDITOR=/bin/true bundle exec "rails credentials:edit --environment $RAILS_ENV"` to generate on-the-fly a new credentials.yml.enc with the given RAILS_MASTER_KEY.

Reply:
But if we add a warning about an insecure master key, the same should be done for django (and any such others?).

Reply:
It will be used out of the box, if the ENV is defined. If there's no key & no ENV defined, it will be generated on the following command run:

    $ bin/rails credentials:edit

Not sure whether it's warning-worthy in that case, though.

WDYT?

Reply:
@jackorp can you create the `.s2i/assemble`?

Reply:
Hmm, not sure we can get away with just adding the key generation and then just calling `/usr/libexec/s2i/assemble`...

By trying to generate the key I am hitting the chicken or the egg problem.
If I try to generate the key before calling the rest of the s2i script, the packages are not yet installed, if I try to generate it as the last thing, it fails on rails asset compilation because it does not have the key and credentials yet.

Reply:
Well, it doesn't seem it contains much logic for rails-related stuff:

https://github.com/sclorg/s2i-ruby-container/blob/master/3.0/s2i/bin/assemble

I think we can run `  bundle exec "rake assets:precompile"` ourselves :) afterwards it will simply result it NOOP.

Reply:
Actually, I'm not sure that's the command we want to run (sorry), this one seems better:
https://github.com/sclorg/s2i-ruby-container/blob/master/3.0/s2i/bin/assemble#L51
I wouldn't worry about any args though (but we can copy the logic whole to be sure).

Reply:
>  I wouldn't worry about any args though (but we can copy the logic whole to be sure).

Sure, sounds good.

Reply:
Opened PR with the assemble script that seems to be working: https://github.com/pvalena/rails-ex/pull/3